lucet-idl/src/rust.rs

```diff
@@ -1,289 +1,460 @@
-        module: &Module,
-        dt: &Named<DataType>,
-        struct_: &StructDataType,
-    ) -> Result<(), IDLError> {
-        let typename = self.define_name(dt);
-
+    fn gen_struct(&mut self, struct_: &StructDatatype) -> Result<(), IDLError> {
         self.w
-            .writeln("#[repr(C)]")?
-            .writeln(format!("pub struct {} {{", typename))?;
+            .writeln("#[repr(C)]")
+            .writeln("#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]")
+            .writeln(format!("pub struct {} {{", struct_.rust_type_name()));
 
         let mut w = self.w.new_block();
-        for m in struct_.members.iter() {
+        for m in struct_.members() {
             w.writeln(format!(
-                "{}: {},",
-                m.name.to_snake_case(),
-                self.get_defined_typename(&m.type_)
-            ))?;
+                "pub {}: {},",
+                m.rust_name(),
+                m.type_().rust_type_name(),
+            ));
+        }
+
+        self.w.writeln("}");
+
+        self.w
+            .writeln(format!("impl {} {{", struct_.rust_type_name()));
+        let mut w = self.w.new_block();
+
+        w.writeln("pub fn validate_bytes(repr: &[u8]) -> bool {");
+        let mut ww = w.new_block();
+        for m in struct_.members() {
+            match m.type_().anti_alias().variant() {
+                DatatypeVariant::Struct(_) | DatatypeVariant::Enum(_) => {
+                    ww.writeln(format!(
+                        "{}::validate_bytes(&repr[{}..{}]) &&",
+                        m.type_().rust_type_name(),
+                        m.offset(),
+                        m.offset() + m.type_().mem_size(),
+                    ));
+                }
+                DatatypeVariant::Atom(_) => {
+                    ww.writeln(format!("// not validating {}", m.type_().rust_type_name()));
+                }
+                DatatypeVariant::Alias(_) => unreachable!("anti-aliased datatype"),
+            }
         }
+        ww.writeln("true"); // to catch the trailing &&
+        w.writeln("}");
 
-        self.w.writeln("}")?.eob()?;
+        self.w.writeln("}"); // end impl
 
-        gen_testcase(&mut self.w, &dt.name.name.to_snake_case(), |w| {
+        gen_testcase(&mut self.w, &struct_.name().to_snake_case(), |w| {
             w.writeln(format!(
                 "assert_eq!({}, ::std::mem::size_of::<super::{}>());",
-                dt.entity.repr_size, typename
-            ))?;
+                struct_.mem_size(),
+                struct_.rust_type_name(),
+            ));
 
-            for m in struct_.members.iter() {
+            for m in struct_.members() {
                 w.writeln(format!(
                     "assert_eq!({}, {{ let base = ::std::ptr::null::<super::{}>(); unsafe {{ (&(*base).{}) as *const _ as usize }} }});",
-                    m.offset, typename, m.name,
-                ))?;
+                    m.offset(), struct_.rust_type_name(), m.rust_name(),
```

Comment:
has this method of checking offsets survived the `offset_of` Discourse?

Reply:
nope its hella UB and should be corrected

Reply:
its fixed.